dotnet/samples/Demos/VectorStoreRAG/DataLoader.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using Microsoft.SemanticKernel.Data;
+using Microsoft.SemanticKernel.Embeddings;
+using UglyToad.PdfPig;
+using UglyToad.PdfPig.Content;
+using UglyToad.PdfPig.DocumentLayoutAnalysis.PageSegmenter;
+
+namespace VectorStoreRAG;
+
+/// <summary>
+/// Class that loads text from a PDF file into a vector store.
+/// </summary>
+/// <typeparam name="TKey">The type of the data model key.</typeparam>
+/// <param name="uniqueKeyGenerator">A function to generate unique keys with.</param>
+/// <param name="vectorStoreRecordCollection">The collection to load the data into.</param>
+/// <param name="textEmbeddingGenerationService">The service to use for generating embeddings from the text.</param>
+internal sealed class DataLoader<TKey>(
+    UniqueKeyGenerator<TKey> uniqueKeyGenerator,
+    IVectorStoreRecordCollection<TKey, TextSnippet<TKey>> vectorStoreRecordCollection,
+    ITextEmbeddingGenerationService textEmbeddingGenerationService) : IDataLoader where TKey : notnull
+{
+    /// <inheritdoc/>
+    public async Task LoadPdf(string pdfPath, CancellationToken cancellationToken)
+    {
+        // Create the collection if it doesn't exist.
+        await vectorStoreRecordCollection.CreateCollectionIfNotExistsAsync(cancellationToken).ConfigureAwait(false);
+
+        // Load the paragraphs from the PDF file and split them into batches.
+        var sectionsText = LoadParagraphs(pdfPath, cancellationToken);
+        var batches = sectionsText.Chunk(10);
+
+        // Process each batch of paragraphs.
+        var sectionNumber = 0;
+        foreach (var batch in batches)
+        {
+            // Map each paragraph to a TextSnippet and generate an embedding for it.
+            var recordTasks = batch.Select(async sectionText => new TextSnippet<TKey>
+            {
```

Comment:
Include the name of the PDF File, so we know which file the snippet comes from. Also include the page number so the citation can point to the specific page the information comes from.

Reply:
Done.